src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs

```diff
@@ -6098,10 +5968,167 @@ private void LoadIndexOfAnyValues(ReadOnlySpan<char> chars)
             int index = list.Count;
             list.Add(IndexOfAnyValues.Create(chars));
 
-            // this._indexOfAnyValues[index]
+            // Logically do _indexOfAnyValues[index], but avoid the bounds check on accessing the array,
+            // and cast to the known derived sealed type to enable devirtualization.
+
+            // Unsafe.As<DerivedIndexOfAnyValues>(Unsafe.Add(ref MemoryMarshal.GetArrayDataReference(this._indexOfAnyValues), index))
             Ldthisfld(s_indexOfAnyValuesArrayField);
-            Ldc(index);
-            _ilg!.Emit(OpCodes.Ldelem_Ref);
+            Call(s_memoryMarshalGetArrayDataReferenceIndexOfAnyValues);
+            Ldc(index * IntPtr.Size);
+            Add();
+            _ilg!.Emit(OpCodes.Ldind_Ref);
+            Call(s_unsafeAs.MakeGenericMethod(list[index].GetType()));
```

Comment:
Did you see a difference with doing the `Unsafe.As` here?
I remember experimenting with doing just that when I initially added these, but it made no difference (it's possible I messed something else up).

@EgorBo would you expect the call to `GetVal` to go directly to the sealed override, and potentially even get inlined in a case like this?

```c#
public class Devirt
{
    public class TestClass
    {
        public virtual int GetVal() => 123;
    }

    public sealed class TestClassSealed : TestClass
    {
        public sealed override int GetVal() => 42;
    }

    private TestClass _field = new TestClassSealed();

    [Benchmark]
    public void GetValUnsafe() => Unsafe.As<TestClass, TestClassSealed>(ref _field).GetVal();
}
```

```assembly
; Devirt.GetValUnsafe()
       sub       rsp,28
       mov       rcx,[rcx+8]
       mov       rax,[rcx]
       mov       rax,[rax+40]
       call      qword ptr [rax+20]
       nop
       add       rsp,28
       ret
; Total bytes of code 24
```

Reply:
@MihaZupan good catch, as per discussion in Discord, it needs a bit of help from JIT and should be a simple fix I hope

Reply:
I really thought it was, but I must have been looking at a different iteration where I had some helper methods declared to actually return the right type.  This devirtualizes:
```C#
using System.Runtime.CompilerServices;

internal class Program
{
    static int Main() => Get().M();

    [MethodImpl(MethodImplOptions.NoInlining)]
    static Derived Get() => new Derived();
}

class Base { public virtual int M() => 1; }
sealed class Derived : Base { public override int M() => 2; }
```
to:
```
G_M000_IG02:                ;; offset=0004H
       FF15CE1A2400         call     [Program:Get():Derived]
       3800                 cmp      byte  ptr [rax], al
       B802000000           mov      eax, 2
```
but neither:
```C#
using System.Runtime.CompilerServices;

internal class Program
{
    static int Main() => Unsafe.As<Derived>(Get()).M();

    [MethodImpl(MethodImplOptions.NoInlining)]
    static Base Get() => new Derived();
}

class Base { public virtual int M() => 1; }
sealed class Derived : Base { public override int M() => 2; }
```
nor:
```C#
using System.Runtime.CompilerServices;

internal class Program
{
    static int Main()
    {
        Derived d = Unsafe.As<Derived>(Get());
        return d.M();
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    static Base Get() => new Derived();
}

class Base { public virtual int M() => 1; }
sealed class Derived : Base { public override int M() => 2; }
```
are able to.

Reply:
I pushed a change that fixes the devirtualization.  As an example, for the regex `@"\w+\d"`, this is what's now generated for the matching routine:
```asm
; Assembly listing for method System.Text.RegularExpressions.CompiledRegexRunner:Regex1_TryMatchAtCurrentPosition(System.Text.RegularExpressions.RegexRunner,System.ReadOnlySpan`1[ushort]):bool
; Emitting BLENDED_CODE for X64 CPU with AVX - Windows
; optimized code
; rsp based frame
; fully interruptible
; No PGO data
; 9 inlinees with PGO data; 26 single block inlinees; 11 inlinees without PGO data

G_M000_IG01:                ;; offset=0000H
       4157                 push     r15
       4156                 push     r14
       4155                 push     r13
       4154                 push     r12
       57                   push     rdi
       56                   push     rsi
       55                   push     rbp
       53                   push     rbx
       4883EC38             sub      rsp, 56
       C5F877               vzeroupper
       488BF1               mov      rsi, rcx

G_M000_IG02:                ;; offset=0016H
       488B3A               mov      rdi, bword ptr [rdx]
       8B5A08               mov      ebx, dword ptr [rdx+08H]
       8B6E4C               mov      ebp, dword ptr [rsi+4CH]
       448BF5               mov      r14d, ebp
       443BF3               cmp      r14d, ebx
       0F87D8020000         ja       G_M000_IG43
       418BCE               mov      ecx, r14d
       4C8D3C4F             lea      r15, bword ptr [rdi+2*rcx]
       448BE3               mov      r12d, ebx
       452BE6               sub      r12d, r14d
       488B4E78             mov      rcx, gword ptr [rsi+78H]
       3809                 cmp      byte  ptr [rcx], cl
       488B4918             mov      rcx, gword ptr [rcx+18H]
       4885C9               test     rcx, rcx
       0F84AC020000         je       G_M000_IG42
       4183FC08             cmp      r12d, 8
       7D11                 jge      SHORT G_M000_IG04

G_M000_IG03:                ;; offset=0051H
       498BD7               mov      rdx, r15
       458BC4               mov      r8d, r12d
       FF1563D71500         call     [System.Buffers.IndexOfAnyAsciiCharValues`1[System.Buffers.IndexOfAnyAsciiSearcher+Default]:IndexOfAnyScalar[System.Buffers.IndexOfAnyAsciiSearcher+Negate](byref,int):int:this]
       448BE8               mov      r13d, eax
       EB1F                 jmp      SHORT G_M000_IG05
                            align    [0 bytes for IG16]

G_M000_IG04:                ;; offset=0062H
       C5F8104108           vmovups  xmm0, xmmword ptr [rcx+08H]
       C5F829442420         vmovaps  xmmword ptr [rsp+20H], xmm0
       498BCF               mov      rcx, r15
       418BD4               mov      edx, r12d
       4C8D442420           lea      r8, [rsp+20H]
       FF155AD71500         call     [System.Buffers.IndexOfAnyAsciiSearcher:IndexOfAnyVectorized[System.Buffers.IndexOfAnyAsciiSearcher+Negate,System.Buffers.IndexOfAnyAsciiSearcher+Default](byref,int,System.Runtime.Intrinsics.Vector128`1[ubyte]):int]
       448BE8               mov      r13d, eax

G_M000_IG05:                ;; offset=0081H
       453BEC               cmp      r13d, r12d
       7366                 jae      SHORT G_M000_IG12

G_M000_IG06:                ;; offset=0086H
       418BCD               mov      ecx, r13d
       6641833C4F7F         cmp      word  ptr [r15+2*rcx], 127
       765B                 jbe      SHORT G_M000_IG12

G_M000_IG07:                ;; offset=0091H
       453BEC               cmp      r13d, r12d
       0F8370020000         jae      G_M000_IG44
       418BCD               mov      ecx, r13d
       410FB70C4F           movzx    rcx, word  ptr [r15+2*rcx]
       48B89877789971010000 mov      rax, 0x17199787798
       8BD1                 mov      edx, ecx
       C1FA03               sar      edx, 3
       83FA10               cmp      edx, 16
       721A                 jb       SHORT G_M000_IG09

G_M000_IG08:                ;; offset=00B6H
       FF15B4551500         call     [System.Globalization.CharUnicodeInfo:GetUnicodeCategoryNoBoundsChecks(uint):int]
       B901000000           mov      ecx, 1
       C4E279F7C9           shlx     ecx, ecx, eax
       F7C13F010400         test     ecx, 0x4013F
       741E                 je       SHORT G_M000_IG12
       EB0E                 jmp      SHORT G_M000_IG10

G_M000_IG09:                ;; offset=00D0H
       8BD2                 mov      edx, edx
       0FB61410             movzx    rdx, byte  ptr [rax+rdx]
       83E107               and      ecx, 7
       0FA3CA               bt       edx, ecx
       730E                 jae      SHORT G_M000_IG12

G_M000_IG10:                ;; offset=00DEH
       41FFC5               inc      r13d
       453BEC               cmp      r13d, r12d
       72AB                 jb       SHORT G_M000_IG07

G_M000_IG11:                ;; offset=00E6H
       41BDFFFFFFFF         mov      r13d, -1

G_M000_IG12:                ;; offset=00ECH
       4585ED               test     r13d, r13d
       450F4CEC             cmovl    r13d, r12d
       4585ED               test     r13d, r13d
       7F11                 jg       SHORT G_M000_IG14

G_M000_IG13:                ;; offset=00F8H
       396E4C               cmp      dword ptr [rsi+4CH], ebp
       0F8DE3010000         jge      G_M000_IG40
       896E4C               mov      dword ptr [rsi+4CH], ebp
       E9DB010000           jmp      G_M000_IG40

G_M000_IG14:                ;; offset=0109H
       453BEC               cmp      r13d, r12d
       0F87F1010000         ja       G_M000_IG43
       418BCD               mov      ecx, r13d
       4D8D3C4F             lea      r15, bword ptr [r15+2*rcx]
       452BE5               sub      r12d, r13d
       438D2C2E             lea      ebp, [r14+r13]
       448BED               mov      r13d, ebp
       418D4601             lea      eax, [r14+01H]
       89442434             mov      dword ptr [rsp+34H], eax
       E90C010000           jmp      G_M000_IG30

G_M000_IG15:                ;; offset=0130H
       413BC5               cmp      eax, r13d
       0F8DAB010000         jge      G_M000_IG40
       442BE8               sub      r13d, eax
       8BC8                 mov      ecx, eax
       418BD5               mov      edx, r13d
       4803CA               add      rcx, rdx
       8BD3                 mov      edx, ebx
       483BCA               cmp      rcx, rdx
       0F87B4010000         ja       G_M000_IG43
       8BC8                 mov      ecx, eax
       4C8D244F             lea      r12, bword ptr [rdi+2*rcx]
       488B4E78             mov      rcx, gword ptr [rsi+78H]
       3809                 cmp      byte  ptr [rcx], cl
       488B4920             mov      rcx, gword ptr [rcx+20H]
       4885C9               test     rcx, rcx
       0F848F010000         je       G_M000_IG42
       4183FD08             cmp      r13d, 8
       7D3D                 jge      SHORT G_M000_IG22
       418D55FF             lea      edx, [r13-01H]
       85D2                 test     edx, edx
       7C2A                 jl       SHORT G_M000_IG19

G_M000_IG16:                ;; offset=0176H
       4C63C2               movsxd   r8, edx
       470FB70444           movzx    r8, word  ptr [r12+2*r8]
       4C8D4918             lea      r9, bword ptr [rcx+18H]
       4181F880000000       cmp      r8d, 128
       7D1C                 jge      SHORT G_M000_IG20

G_M000_IG17:                ;; offset=018BH
       458BD0               mov      r10d, r8d
       41C1FA05             sar      r10d, 5
       478B0C91             mov      r9d, dword ptr [r9+4*r10]
       450FA3C1             bt       r9d, r8d
       730B                 jae      SHORT G_M000_IG20

G_M000_IG18:                ;; offset=019CH
       FFCA                 dec      edx
       79D6                 jns      SHORT G_M000_IG16

G_M000_IG19:                ;; offset=01A0H
       BDFFFFFFFF           mov      ebp, -1
       EB02                 jmp      SHORT G_M000_IG21

G_M000_IG20:                ;; offset=01A7H
       8BEA                 mov      ebp, edx

G_M000_IG21:                ;; offset=01A9H
       EB1E                 jmp      SHORT G_M000_IG23

G_M000_IG22:                ;; offset=01ABH
       C5F8104108           vmovups  xmm0, xmmword ptr [rcx+08H]
       C5F829442420         vmovaps  xmmword ptr [rsp+20H], xmm0
       498BCC               mov      rcx, r12
       418BD5               mov      edx, r13d
       4C8D442420           lea      r8, [rsp+20H]
       FF1571D61500         call     [System.Buffers.IndexOfAnyAsciiSearcher:LastIndexOfAnyVectorized[System.Buffers.IndexOfAnyAsciiSearcher+Negate,System.Buffers.IndexOfAnyAsciiSearcher+Ssse3AndWasmHandleZeroInNeedle](byref,int,System.Runtime.Intrinsics.Vector128`1[ubyte]):int]
       8BE8                 mov      ebp, eax

G_M000_IG23:                ;; offset=01C9H
       413BED               cmp      ebp, r13d
       7343                 jae      SHORT G_M000_IG29
       8BCD                 mov      ecx, ebp
       6641833C4C7F         cmp      word  ptr [r12+2*rcx], 127
       7639                 jbe      SHORT G_M000_IG29

G_M000_IG24:                ;; offset=01D8H
       413BED               cmp      ebp, r13d
       0F8329010000         jae      G_M000_IG44
       8BCD                 mov      ecx, ebp
       410FB70C4C           movzx    rcx, word  ptr [r12+2*rcx]
       81F900010000         cmp      ecx, 256
       730A                 jae      SHORT G_M000_IG26

G_M000_IG25:                ;; offset=01F0H
       83C1D0               add      ecx, -48
       83F909               cmp      ecx, 9
       7619                 jbe      SHORT G_M000_IG29
       EB0B                 jmp      SHORT G_M000_IG27

G_M000_IG26:                ;; offset=01FAH
       FF1570541500         call     [System.Globalization.CharUnicodeInfo:GetUnicodeCategoryNoBoundsChecks(uint):int]
       83F808               cmp      eax, 8
       740C                 je       SHORT G_M000_IG29

G_M000_IG27:                ;; offset=0205H
       FFCD                 dec      ebp
       413BED               cmp      ebp, r13d
       72CC                 jb       SHORT G_M000_IG24

G_M000_IG28:                ;; offset=020CH
       BDFFFFFFFF           mov      ebp, -1

G_M000_IG29:                ;; offset=0211H
       448BED               mov      r13d, ebp
       4585ED               test     r13d, r13d
       0F8CC7000000         jl       G_M000_IG40
       8B442434             mov      eax, dword ptr [rsp+34H]
       4403E8               add      r13d, eax
       418BED               mov      ebp, r13d
       3BEB                 cmp      ebp, ebx
       0F87D4000000         ja       G_M000_IG43
       418BCD               mov      ecx, r13d
       4C8D3C4F             lea      r15, bword ptr [rdi+2*rcx]
       448BE3               mov      r12d, ebx
       452BE5               sub      r12d, r13d

G_M000_IG30:                ;; offset=023CH
       396E4C               cmp      dword ptr [rsi+4CH], ebp
       7D03                 jge      SHORT G_M000_IG32

G_M000_IG31:                ;; offset=0241H
       896E4C               mov      dword ptr [rsi+4CH], ebp

G_M000_IG32:                ;; offset=0244H
       4585E4               test     r12d, r12d
       0F86E3FEFFFF         jbe      G_M000_IG15
       410FB70F             movzx    rcx, word  ptr [r15]
       81F900010000         cmp      ecx, 256
       730E                 jae      SHORT G_M000_IG35

G_M000_IG33:                ;; offset=0259H
       83C1D0               add      ecx, -48
       83F909               cmp      ecx, 9
       0F87CBFEFFFF         ja       G_M000_IG15

G_M000_IG34:                ;; offset=0265H
       EB13                 jmp      SHORT G_M000_IG36

G_M000_IG35:                ;; offset=0267H
       FF1503541500         call     [System.Globalization.CharUnicodeInfo:GetUnicodeCategoryNoBoundsChecks(uint):int]
       83F808               cmp      eax, 8
       8B442434             mov      eax, dword ptr [rsp+34H]
       0F85B6FEFFFF         jne      G_M000_IG15

G_M000_IG36:                ;; offset=027AH
       FFC5                 inc      ebp
       896E4C               mov      dword ptr [rsi+4CH], ebp
       8BFD                 mov      edi, ebp
       413BFE               cmp      edi, r14d
       7D08                 jge      SHORT G_M000_IG37
       8BC7                 mov      eax, edi
       418BFE               mov      edi, r14d
       448BF0               mov      r14d, eax

G_M000_IG37:                ;; offset=028EH
       837E5800             cmp      dword ptr [rsi+58H], 0
       7509                 jne      SHORT G_M000_IG38
       488BCE               mov      rcx, rsi
       FF159BE90700         call     [System.Text.RegularExpressions.RegexRunner:DoubleCrawl():this]

G_M000_IG38:                ;; offset=029DH
       4C8B4E20             mov      r9, gword ptr [rsi+20H]
       8B4E58               mov      ecx, dword ptr [rsi+58H]
       FFC9                 dec      ecx
       894E58               mov      dword ptr [rsi+58H], ecx
       413B4908             cmp      ecx, dword ptr [r9+08H]
       735B                 jae      SHORT G_M000_IG44
       4533C0               xor      r8d, r8d
       4589448910           mov      dword ptr [r9+4*rcx+10H], r8d
       448BCF               mov      r9d, edi
       452BCE               sub      r9d, r14d
       488B4E28             mov      rcx, gword ptr [rsi+28H]
       458BC6               mov      r8d, r14d
       33D2                 xor      edx, edx
       3909                 cmp      dword ptr [rcx], ecx
       FF1562881500         call     [System.Text.RegularExpressions.Match:AddMatch(int,int,int):this]
       B801000000           mov      eax, 1

G_M000_IG39:                ;; offset=02D3H
       4883C438             add      rsp, 56
       5B                   pop      rbx
       5D                   pop      rbp
       5E                   pop      rsi
       5F                   pop      rdi
       415C                 pop      r12
       415D                 pop      r13
       415E                 pop      r14
       415F                 pop      r15
       C3                   ret

G_M000_IG40:                ;; offset=02E4H
       33C0                 xor      eax, eax

G_M000_IG41:                ;; offset=02E6H
       4883C438             add      rsp, 56
       5B                   pop      rbx
       5D                   pop      rbp
       5E                   pop      rsi
       5F                   pop      rdi
       415C                 pop      r12
       415D                 pop      r13
       415E                 pop      r14
       415F                 pop      r15
       C3                   ret

G_M000_IG42:                ;; offset=02F7H
       B906000000           mov      ecx, 6
       FF15E6D00300         call     [System.ThrowHelper:ThrowArgumentNullException(int)]
       CC                   int3

G_M000_IG43:                ;; offset=0303H
       FF15C7CD0300         call     [System.ThrowHelper:ThrowArgumentOutOfRangeException()]
       CC                   int3

G_M000_IG44:                ;; offset=030AH
       E821B7A45F           call     CORINFO_HELP_RNGCHKFAIL
       CC                   int3

; Total bytes of code 784
```